packages/dev/core/src/Meshes/GaussianSplatting/gaussianSplattingMesh.ts

```diff
@@ -269,259 +514,271 @@ export class GaussianSplattingMesh extends Mesh {
+                    break;
+                case PLYValue.F_DC_2:
+                    rgba[2] = (0.5 + GaussianSplattingMesh._SH_C0 * value) * 255;
+                    break;
+                case PLYValue.F_DC_3:
+                    rgba[3] = (0.5 + GaussianSplattingMesh._SH_C0 * value) * 255;
+                    break;
+                case PLYValue.OPACITY:
+                    rgba[3] = (1 / (1 + Math.exp(-value))) * 255;
+                    break;
+                case PLYValue.ROT_0:
+                    r0 = value;
+                    break;
+                case PLYValue.ROT_1:
+                    r1 = value;
+                    break;
+                case PLYValue.ROT_2:
+                    r2 = value;
+                    break;
+                case PLYValue.ROT_3:
+                    r3 = value;
+                    break;
+            }
         }
 
-        return buffer;
+        q.set(r1, r2, r3, r0);
+        q.normalize();
+        rot[0] = q.w * 128 + 128;
+        rot[1] = q.x * 128 + 128;
+        rot[2] = q.y * 128 + 128;
+        rot[3] = q.z * 128 + 128;
+        offset.value += header.rowVertexLength;
+    }
+
+    /**
+     * Converts a .ply data array buffer to splat
+     * if data array buffer is not ply, returns the original buffer
+     * @param data the .ply data to load
+     * @returns the loaded splat buffer
+     */
+    public static ConvertPLYToSplat(data: ArrayBuffer): ArrayBuffer {
+        const header = GaussianSplattingMesh.ParseHeader(data);
+        if (!header) {
+            return data;
+        }
+
+        const offset = { value: 0 };
+        const compressedChunks = GaussianSplattingMesh._GetCompressedChunks(header, offset);
+
+        for (let i = 0; i < header.vertexCount; i++) {
+            GaussianSplattingMesh._GetSplat(header, i, compressedChunks, offset);
+        }
+
+        return header.buffer;
+    }
+
+    /**
+     * Converts a .ply data array buffer to splat
+     * if data array buffer is not ply, returns the original buffer
+     * @param data the .ply data to load
+     * @returns the loaded splat buffer
+     */
+    public static async ConvertPLYToSplatAsync(data: ArrayBuffer): Promise<ArrayBuffer> {
+        return new Promise(async (resolve) => {
+            const header = GaussianSplattingMesh.ParseHeader(data);
+            if (!header) {
+                resolve(data);
+                return;
+            }
+
+            const offset = { value: 0 };
+            const compressedChunks = GaussianSplattingMesh._GetCompressedChunks(header, offset);
+
+            for (let i = 0; i < header.vertexCount; i++) {
+                GaussianSplattingMesh._GetSplat(header, i, compressedChunks, offset);
+
+                // todo: this number should be part of the loading options
+                if (i % 30000 === 0) {
+                    await new Promise((resolve) => setTimeout(resolve, 1));
+                }

```

Comment:
I assume this is what you were referring to as a "coroutine like mechanism" (e.g. just yielding occasionally)? If so, also note that we have some helper functions with more formal "coroutine" support. It uses iterators/yield, so the underlying "coroutine" function can be run synchronously or asynchronously. In the async case, you can specify a scheduler, and there is a default scheduler that is time based (25ms by default). You can see an example with `MergeMeshes` and `MergeMeshesAsync` in mesh.ts.

Reply:
Fixed